pilot/pkg/model/push_context.go

```diff
@@ -310,15 +312,50 @@ func (ps *PushContext) UpdateMetrics() {
 }
 
 // Services returns the list of services that are visible to a Proxy in a given config namespace
-func (ps *PushContext) Services(proxy *Proxy) []*Service {
+func (ps *PushContext) Services(p *Proxy) []*Service {
 	// TODO: use network scopes here, and return services in the proxy namespace if
 	// configured to use current_namespace as the scope and proxy is not nil
-	//if proxy == nil {
-	//	return ps.allServices
-	//}
+	if p == nil {
+		return ps.allServices
+	}
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+	if p.serviceDependencies != nil {
+		return p.serviceDependencies
+	}
 	return ps.allServices
 }
 
+// OnConnect is called when a node connects. Will update per-node data.
+func (ps *PushContext) OnConnect(p *Proxy) {
+	// For now Router (Gateway) is not using the isolation - the Gateway already has explicit
+	// bindings.
+	if pilot.NetworkScopes != "" && p.Type == Sidecar {
+		// Add global namespaces. This may be loaded from mesh config ( after the API is stable and
+		// reviewed ), or from an env variable.
+		adminNs := strings.Split(pilot.NetworkScopes, ",")
```

Comment:
Please revert this logic and parse the list of namespaces from the mesh config map. I definitely do not want to supply the shared namespaces via an environment variable. @hzxuzhonghu 's PR extracts the info nicely from the mesh config and adds it. Even if other parts of his PR are not optimized, atleast please use the code related to this logic. OR leave this function empty and let him fill it.